### PR TITLE
TF-942 Auto generate missing default mailboxes in client side

### PR DESCRIPTION
**Pull request by dab246**
#942 

### Related

- [https://github.com/linagora/jmap-dart-client/pull/38](https://github.com/linagora/jmap-dart-client/pull/38)
- [https://github.com/linagora/jmap-dart-client/pull/39](https://github.com/linagora/jmap-dart-client/pull/39)

### Resolved

- For `jmap.linagora.com`

https://user-images.githubusercontent.com/80730648/191917891-b7144ed6-ec3c-429b-821a-118bc1308a55.mov

- For `api.fastmail.com` has mailbox with name `Outbox`

https://user-images.githubusercontent.com/80730648/191918237-9de88a60-34aa-49d2-a931-558598cb7c74.mov

- For `api.fastmail.com` has not mailbox with name `Outbox`

https://user-images.githubusercontent.com/80730648/191918403-9a2cd5af-5d87-4c4f-949e-288c54de05f7.mov

### CODE REVIEW

File: lib/features/composer/domain/usecases/send_email_interactor.dart [hunk 1]
```diff
@@ -1,39 +1,67 @@
...
   final MailboxRepository _mailboxRepository;
+  final Uuid _uuid;
 
-  SendEmailInteractor(this._emailRepository, this._mailboxRepository);
+  SendEmailInteractor(this._emailRepository, this._mailboxRepository, this._uuid);
 
   Stream<Either<Failure, Success>> execute(AccountId accountId, EmailRequest emailRequest) async* {
     try {
       yield Right<Failure, Success>(SendingEmailState());
+      final newOutboxMailboxId = emailRequest.outboxMailboxId ?? await _autoGenerateMissingOutboxMailbox(accountId);
```
Comment (hoangdat) [inline lib/features/composer/domain/usecases/send_email_interactor.dart:26]: we can not `create mailbox -> then send email` in the same Jmap request?

**Comment by hoangdat** [comment]
hi @chibenwa @Arsnael How we can create mailbox and send email in the same JMAP request. I tried, but the `json pointer` too hard to me implement that.
### Use-case
```
As a user, who don't have Outbox folder
I want to send email 
```

### Proposal
- Create `Outbox` folder if it is not existed (`c1`)
- Set Email with created mailbox in `c1` (`c2`)
- Set EmailSubmission and move Email in `c2` to Sent folder

Please help me to compose the JMAP request with it if we can. (We are implementing it in 2 separate JMAP requests)

File: lib/features/composer/domain/usecases/send_email_interactor.dart (hunk 1 above)
Comment (dab246) [reply lib/features/composer/domain/usecases/send_email_interactor.dart:26]: I tried and it's too hard to take a `property` like `id` in the returned result and reassign it to an object created from the client? If you have a solution for this please suggest me. Thanks

**Comment by Arsnael** [comment]
@hoangdat @dab246 

Should work with something like that: https://github.com/Arsnael/james-project/commit/622748f41d0a4c0f2908411c5b5b29b2ff1cb39c

If not tell me what's wrong and we can look deeper into this

**Comment by dab246** [comment]
> @hoangdat @dab246
> 
> Should work with something like that: [Arsnael/james-project@622748f](https://github.com/Arsnael/james-project/commit/622748f41d0a4c0f2908411c5b5b29b2ff1cb39c)
> 
> If not tell me what's wrong and we can look deeper into this
> 
> This being the answer I get from the server:
> 
> ```
> {
>     "sessionState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
>     "methodResponses": [
>         [
>             "Mailbox/set",
>             {
>                 "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
>                 "oldState": "bda01c0b-dcbe-4ae1-b70b-e770949d0ec1",
>                 "created": {
>                     "C42": {
>                         "totalThreads": 0,
>                         "isSubscribed": true,
>                         "totalEmails": 0,
>                         "unreadThreads": 0,
>                         "unreadEmails": 0,
>                         "sortOrder": 1000,
>                         "myRights": {
>                             "mayReadItems": true,
>                             "mayAddItems": true,
>                             "mayRemoveItems": true,
>                             "maySetSeen": true,
>                             "maySetKeywords": true,
>                             "mayCreateChild": true,
>                             "mayRename": true,
>                             "mayDelete": true,
>                             "maySubmit": true
>                         },
>                         "id": "3"
>                     }
>                 },
>                 "newState": "6af50957-1157-4e72-a44d-eb87c9dd3921"
>             },
>             "c0"
>         ],
>         [
>             "Email/set",
>             {
>                 "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
>                 "oldState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
>                 "newState": "55be4432-7ece-466a-8d1b-5b95e38e4768",
>                 "created": {
>                     "e1526": {
>                         "id": "1",
>                         "blobId": "1",
>                         "threadId": "1",
>                         "size": 383
>                     }
>                 }
>             },
>             "c1"
>         ],
>         [
>             "EmailSubmission/set",
>             {
>                 "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
>                 "newState": "2c9f1b12-b35a-43e6-9af2-0106fb53a943",
>                 "created": {
>                     "k1490": "1f8ace54-0ec1-44c5-bd2d-98f8358d333c"
>                 }
>             },
>             "c2"
>         ],
>         [
>             "Email/set",
>             {
>                 "accountId": "29883977c13473ae7cb7678ef767cbfbaffc8a44a6e463d971d23a65c1dc4af6",
>                 "oldState": "55be4432-7ece-466a-8d1b-5b95e38e4768",
>                 "newState": "c820f03a-d832-4b9a-9f0a-ece47af5630a",
>                 "updated": {
>                     "1": null
>                 }
>             },
>             "c2"
>         ]
>     ]
> }
> ```

@Arsnael Cool, Thanks u.

File: lib/features/email/data/network/email_api.dart [hunk 2]
```diff
@@ -76,39 +80,61 @@ class EmailAPI {
...
+
+    if (mailboxRequest != null) {
+      final setMailboxMethod = SetMailboxMethod(accountId)
+        ..addCreate(
+            mailboxRequest.creationId,
+            Mailbox(name: mailboxRequest.newName, parentId: mailboxRequest.parentId));
+
+      requestBuilder.invocation(setMailboxMethod);
+
+      outboxMailboxId = MailboxId(ReferenceId(
```
Comment (hoangdat) [inline lib/features/email/data/network/email_api.dart:101]: `outboxCreationReferenceId`

**Review by hoangdat — APPROVED** [review]

**Pull request merged**